### PR TITLE
Feature/chat

### LINKED ISSUE #3: Implement a lobby-wide chat

### PR DESCRIPTION
**Pull request by Tejpbit**
Fully fledged chat. resolves #3 
I currently know of two debatable flaws:

The `msg` command built in can send message to a chat the client hasn't joined.
For example:
`/msg foo Lorem ipsum dolor sit amet`
will send "Lorem ipsum dolor sit amet" to the channel `foo`. regardless if the client is in that chat or not.
Whether we want it like this or not i don't know. It's an easy fix if we only want the clients to be able to send chat messages to chats they've joined but I'll leave it like this for now.

The other flaw happen when two players have the same name.
Say two clients both have chosen the name `apa` and they have joined the channel `foo`.
If either one of them now decides to `/leave` the chat then the other will get kicked out as well.

### CODE REVIEW

**Comment by Kalior** [comment]
What is the reason for allowing all clients to send messages to channels they haven't joined?

**Comment by Tejpbit** [comment]
I used it for debugging and decided not to remove the code. I don't know if it's good for anything when running live.

**Comment by Kalior** [comment]
What are the available commands? Would like to test them. :speech_balloon:

**Comment by Tejpbit** [comment]
I should implement a `/help` command

The commands are:
`/join`
`/leave`
`/msg`

**Comment by Kalior** [comment]
All right, I have a couple of general comments:
- Channel names seems to be case sensitive, is this intentional? It could be confusing.
- Maybe the channels with a player who changes name should notify that the player changed name?
- Would be nice to not have tabs stack on top of each other when there are many (don't know what is supported in bootstrap) screen02
- Also this: 
  screen03

File: src/LobbyServer.hs [hunk 1]
```diff
@@ -353,13 +357,47 @@ joinChat remoteClientList remoteChatList chatName = do
...
+announceChatJoin remoteClientList remoteChatList chatName = do
+  clientList <- remoteClientList >>= liftIO . CC.readMVar
+  chatList <- remoteChatList >>= liftIO . CC.readMVar
+  case chatName `lookup` chatList of
+    Nothing      -> return ()
+    Just channel -> do
+      sid <- getSessionID
+      case sid `lookupClientEntry` clientList of
+        Nothing     -> return ()
+        Just client -> liftIO $ CC.writeChan channel $ ChatAnnounceJoin $ name client
```
Comment (Kalior) [inline src/LobbyServer.hs]: Comment `announceChatJoin` please. :book:

File: src/LobbyServer.hs [hunk 2]
```diff
@@ -353,13 +357,47 @@ joinChat remoteClientList remoteChatList chatName = do
...
+
+-- | Called by client to leave the named Chat
+leaveChat :: Server ConcurrentClientList -> String -> Server ()
+leaveChat remoteClientList chatName = do
+  sid <- getSessionID
```
Comment (Kalior) [inline src/LobbyServer.hs:376]: Perhaps mention what the String does

File: src/LobbyServer.hs [hunk 3]
```diff
@@ -353,13 +357,47 @@ joinChat remoteClientList remoteChatList chatName = do
...
+    Nothing     -> return ()
+    Just client ->
+      case chatName `lookup` chats client of
+        Nothing      -> return ()
+        Just channel -> do
+          liftIO $ do
+            CC.writeChan channel $ ChatAnnounceLeave $ name client
+            CC.modifyMVar_ concurrentClientList $ \clientList ->
+              return $ updateListElem (\c -> c {chats =
+                deleteBy (\(cName1,_) (cName2,_) -> cName1 == cName2) (chatName, channel) $ chats client -- ((chatName ==).). fst) ??
```
Comment (Kalior) [inline src/LobbyServer.hs]: Remove the comment?

File: src/LobbyServer.hs [hunk 4]
```diff
@@ -406,3 +444,14 @@ getClientName remoteClientList = do
   clientList <- liftIO $ CC.readMVar concurrentClientList
   let client = find ((sid ==) . sessionID) clientList
   return $ name $ fromJust client
+
+getJoinedChats :: Server ConcurrentClientList -> Server [String]
+getJoinedChats remoteClientList = do
```
Comment (Kalior) [inline src/LobbyServer.hs:444]: Comment this function? What is returned? I assume the name of the chats?

File: src/Views/Chat.hs [hunk 5]
```diff
@@ -31,82 +35,223 @@ createChatDOMInParent api parentDiv = do
...
+createChatTabsHeader =
+  newElem "ul" `with`
+    [
+      attr "id" =: "chat-tabs",
+      attr "class" =: "nav nav-tabs",
+      attr "role" =: "tablist"
+    ]
+
+addNewChat :: LobbyAPI -> String -> Client ()
+addNewChat api chatName = do
```
Comment (Kalior) [inline src/Views/Chat.hs]: Comment this function please

File: src/Views/Chat.hs [hunk 6]
```diff
@@ -31,82 +35,223 @@ createChatDOMInParent api parentDiv = do
...
+        attr "type" =: "text",
+        attr "cols" =: "60"
+
+      ]
+    onEvent inputField KeyPress $ \13 -> handleChatInput api chatName
+    appendChild inputFieldContainer inputField
+    return ()
+
+deleteChat :: String -> Client ()
+deleteChat chatName = do
```
Comment (Kalior) [inline src/Views/Chat.hs]: Comment this function please

File: src/Views/Chat.hs [hunk 7]
```diff
@@ -31,82 +35,223 @@ createChatDOMInParent api parentDiv = do
...
+setActiveChat chatName = do
+  "chat-tabs" `withElem` \chatTabs -> do
+    setClassOnChildren chatTabs ""
+    maybeChatTab <- elemById $ "chat-tab-" ++ chatName
+    setClassOnMaybeDOMElem maybeChatTab "active"
+  "chat-container" `withElem` \chatContainer -> do
+    setClassOnChildren chatContainer "hide"
+    maybeChatContainer <- elemById $ "chat-container-" ++ chatName
+    setClassOnMaybeDOMElem maybeChatContainer ""
+  "input-container" `withElem` \inputs -> do
```
Comment (Kalior) [inline src/Views/Chat.hs]: You sure this is a safe use of withElem?

File: src/Views/Chat.hs [hunk 8]
```diff
@@ -31,82 +35,223 @@ createChatDOMInParent api parentDiv = do
...
+      setActiveChat chatName
+    else do
+      onServer $ joinChat api <.> chatName
+      addNewChat api chatName
+      fork $ listenForChatMessages api chatName $ chatMessageCallback api chatName
+      setActiveChat chatName
+  return ()
+
+clientLeaveChat :: LobbyAPI -> String -> Client ()
+clientLeaveChat api chatName =
```
Comment (Kalior) [inline src/Views/Chat.hs]: Comment please

**Comment by Kalior** [comment]
Why does it kick multiple players with the same name if one leaves? On the server side this would be easy to fix since the SessionID can get the correct Client who should leave. On the Client side the client can ask for it's chats when it sees someone with the same name has left and check if it still has that channel there or not. Maybe?

File: src/Views/Chat.hs (hunk 7 above)
Comment (Tejpbit) [inline src/Views/Chat.hs]: Unless the user changes the id's in the browsers developer mode then I think this should be safe.

They are however magic constants, I'll extract them...

**Comment by Tejpbit** [comment]
- No longer "kick" all players with same name when one of them leaves.
- Chats are now case insensitive. The first client who joins a chat determines the casing.
- All chats a client has joined are notified when the client changes nick or disconnects.
- Chat tabs are now scrollable. So no fix is needed for long chat names. However long chat names look awful and maybe they should be truncated to some maximum number of characters?

File: src/LobbyServer.hs [hunk 9]
```diff
@@ -55,6 +58,14 @@ connect remoteClientList remoteChats name = do
 -- |Disconnect client from server.
 disconnect :: LobbyState -> SessionID -> Server()
 disconnect (clientList, games, chats) sid = do
+  cs <- clientList >>= liftIO . CC.readMVar
+  maybe
+    (return ())
+    (\c -> do
+      notifyClientChats clientList $ name c ++ " disconnected"
+      return ())
+    $ sid `lookupClientEntry` cs
```
Comment (Kalior) [inline src/LobbyServer.hs:78]: If you `return ()' in both the cases, why not do it afterwards?

File: src/LobbyServer.hs [hunk 10]
```diff
@@ -235,9 +252,23 @@ changeNickName remoteClientList remoteGames newName = do
...
 
+-- | Sends a server notification to all chats the client has joined
+notifyClientChats :: Server ConcurrentClientList -> String -> Server ()
+notifyClientChats remoteClients notification = do
+  clientList <- remoteClients >>= liftIO . CC.readMVar
+  sid <- getSessionID
+  maybe
+    (liftIO . print $ "notifyClientChats > Could not find sid in connected clients")
+    (liftIO . mapM_ (flip CC.writeChan (ChatMessage "SERVER" notification) . snd) . chats)
+    $ sid `lookupClientEntry` clientList
```
Comment (Kalior) [inline src/LobbyServer.hs:245]: If you `liftIO` in both cases, why not do it before `maybe`?

**Comment by Kalior** [comment]
I think you gone broke it?
screen04

Added some minor comments in the code, otherwise looks fine :dancers:

File: src/LobbyServer.hs (hunk 9 above)
Comment (Kalior) [inline src/LobbyServer.hs:78]: Did you remove the redundant do and return () in the Just case?

**Pull request merged**